repository/state/state.go

```diff
@@ -141,37 +129,256 @@ func (st *State) getOrCreateIdForChecksum(checksum objects.Checksum) uint64 {
 	return newID
 }
 
-func NewFromBytes(serialized []byte) (*State, error) {
-	if len(serialized) < 4 {
-		return nil, fmt.Errorf("invalid state data")
+func (st *State) SerializeStream(w io.Writer) error {
+	writeUint64 := func(value uint64) error {
+		buf := make([]byte, 8)
+		binary.LittleEndian.PutUint64(buf, value)
+		_, err := w.Write(buf)
+		return err
+	}
+
+	writeUint32 := func(value uint32) error {
+		buf := make([]byte, 4)
+		binary.LittleEndian.PutUint32(buf, value)
+		_, err := w.Write(buf)
+		return err
+	}
+
+	writeLocation := func(loc Location) error {
+		if err := writeUint64(loc.Packfile); err != nil {
+			return err
+		}
+		if err := writeUint32(loc.Offset); err != nil {
+			return err
+		}
+		return writeUint32(loc.Length)
+	}
+
+	// Serialize Metadata
+	if err := writeUint32(st.Metadata.Version); err != nil {
+		return fmt.Errorf("failed to write version: %w", err)
+	}
+	timestamp := st.Metadata.Timestamp.UnixNano()
+	if err := writeUint64(uint64(timestamp)); err != nil {
+		return fmt.Errorf("failed to write timestamp: %w", err)
+	}
+	if st.Metadata.Aggregate {
+		if _, err := w.Write([]byte{1}); err != nil {
+			return fmt.Errorf("failed to write aggregate flag: %w", err)
+		}
+	} else {
+		if _, err := w.Write([]byte{0}); err != nil {
+			return fmt.Errorf("failed to write aggregate flag: %w", err)
+		}
+	}
+	if err := writeUint64(uint64(len(st.Metadata.Extends))); err != nil {
+		return fmt.Errorf("failed to write extends length: %w", err)
+	}
+	for _, checksum := range st.Metadata.Extends {
+		if _, err := w.Write(checksum[:]); err != nil {
+			return fmt.Errorf("failed to write checksum: %w", err)
+		}
+	}
+
+	if err := serializeMapping(w, st.DeletedSnapshots, func(key uint64) error {
+		return writeUint64(key)
+	}, func(value time.Time) error {
```

Comment:
nit: could pass writeUint64 directly instead of wrapping it.

Reply:
fixed